lib/improver/spotdata/site_data.py

```diff
@@ -0,0 +1,302 @@
+        self.altitudes = np.full(n_sites, np.nan)
+        for i_site, site in enumerate(site_data):
+            if 'altitude' in site.keys():
+                self.altitudes[i_site] = site['altitude']
+
+        # If no site ids are provided, use index of read order.
+        self.site_ids = np.full(n_sites, np.nan)
+        for i_site, site in enumerate(site_data):
+            if 'bestdata_id' in site.keys():
+                self.site_ids[i_site] = site['bestdata_id']
+        # If unset use key LATLONALTI
+        # e.g. lat=50, lon=10, altitude=0 --> 0500100000
+        unset_indices, = np.where(np.isnan(self.site_ids))
+        for i_site in unset_indices:
+            self.site_ids[i_site] = '{:03d}{:03d}{:04d}'.format(
+                int(self.latitudes[i_site]+180),
+                int(self.longitudes[i_site]+180),
+                int(np.nanmax([self.altitudes[i_site], 0.])))
+
+        self.utc_offsets = np.full(n_sites, np.nan)
+        for i_site, site in enumerate(site_data):
+            if 'gmtoffset' in site.keys():
+                self.utc_offsets[i_site] = site['gmtoffset']
+            elif 'utcoffset' in site.keys():
+                self.utc_offsets[i_site] = site['utcoffset']
+            elif 'utc_offset' in site.keys():
+                self.utc_offsets[i_site] = site['utc_offset']
+
+            # If it's not been set, set it with the longitude based method.
+            if np.isnan(self.utc_offsets[i_site]):
+                self.utc_offsets[i_site], = set_utc_offset([site['longitude']])
+
+        return self.construct_site_dictionary()
+
+    def runtime_list(self, site_properties):
+        """
+        Use data provided on the command line/controlling suite at runtime.
+
+        Args:
+        -----
+        site_properties : dict
+            Contains:
+
+            latitudes/longitudes : lists of floats
+                Lists of coordinates at which to extract data, these define the
+                positions of the SpotData sites.
+
+            altitudes : list of floats
+                List of altitudes that can optionally be used in defining sites
+                on the fly. If unset the altitudes will be assumed as
+                equivalent to which ever neighbouring grid point is used as a
+                data source.
+
+            site_ids : list of integers
+                Optional integer site_ids that will be used as keys for the
+                site information in the returned site data dictionary. If unset
+                a simple counting index is used.
+
+        Returns:
+        --------
+        sites : dict
+            Restructured dictionary of site data.
+
+        Raises:
+        -------
+        ValueError : If runtime defined lists of site properties are of unequal
+        length, suggesting an error with the input (e.g. a site is missing
+        some required property).
+
+        """
+        n_sites = site_properties['latitudes'].__len__()
+
```

Comment:
I've rejigged things to use a single `parse_input` function for both command line and json. This is what you were suggesting originally, and I'm not sure why I found it confusing or difficult before. Anyway, this particular issue is no longer relevant.